Sources/_Datadog_Private/ObjcAppLaunchHandler.m

```diff
@@ -4,60 +4,94 @@
-                                                     name:UIApplicationDidBecomeActiveNotification
-                                                   object:nil
-         ];
+    if (self = [super init]) {
+        _frameworkLoadTime = frameworkLoadTime;
     }
     return self;
 }
 
-- (void)handleDidBecomeActiveNotification {
-    if (self.timeToApplicationBecomeActive > -1) { // sanity check & extra safety
-        return;
-    }
+- (void)markEndOfUIApplicationMain {
+    _applicationMainEndTime = CFAbsoluteTimeGetCurrent();
+}
 
-    NSDate *time = [NSDate new];
-    NSTimeInterval duration = [time timeIntervalSinceDate:self.frameworkLoadTime];
-    self.timeToApplicationBecomeActive = duration;
-    [[NSNotificationCenter defaultCenter] removeObserver:self];
+- (void)scheduleUIApplicationDidFinishLaunchingMarking {
+    id __block token = [NSNotificationCenter.defaultCenter
+                        addObserverForName:UIApplicationDidFinishLaunchingNotification
+                        object:nil
+                        queue:NSOperationQueue.mainQueue
+                        usingBlock:^(NSNotification *_){
+        [self markUIApplicationDidFinishLaunching];
+        [NSNotificationCenter.defaultCenter removeObserver:token];
+    }];
 }
 
-@end
+- (void)markUIApplicationDidFinishLaunching {
+    NSTimeInterval now = CFAbsoluteTimeGetCurrent();
 
+    [self queryProcessStartTime];
 
-@implementation ObjcAppLaunchHandler: NSObject
+    TimeToApplicationFinishedLaunching = now - _processStartTime;
+    TimeToLoadApplicationRootUI = now - _applicationMainEndTime;
+}
 
-static AppLaunchTimer *_timer;
+- (void)queryProcessStartTime {
+    // Query the current process' start time:
+    // https://www.freebsd.org/cgi/man.cgi?sysctl(3)
+    // https://github.com/darwin-on-arm/xnu/blob/707bfdc4e9a46e3612e53994fffc64542d3f7e72/bsd/sys/sysctl.h#L681
+    // https://github.com/darwin-on-arm/xnu/blob/707bfdc4e9a46e3612e53994fffc64542d3f7e72/bsd/sys/proc.h#L97
 
-+ (void)load {
-    // This is called at the `_Datadog_Private` load time, keep the work minimal
-    _timer = [AppLaunchTimer new];
-}
+    struct kinfo_proc kip;
+    size_t kipSize = sizeof(kip);
+    int mib[4] = {CTL_KERN, KERN_PROC, KERN_PROC_PID, getpid()};
+    int res = sysctl(mib, 4, &kip, &kipSize, NULL, 0);
 
-+ (NSTimeInterval)launchTime {
-    return _timer.timeToApplicationBecomeActive;
+    if (res == 0) {
+        // The process' start time is provided relative to 1 Jan 1970
+        struct timeval startTime = kip.kp_proc.p_starttime;
+        _processStartTime = startTime.tv_sec + startTime.tv_usec / USEC_PER_SEC;
+        // Convert to time since 1 Jan 2001 to align with CFAbsoluteTimeGetCurrent()
+        _processStartTime -= kCFAbsoluteTimeIntervalSince1970;
+    } else {
+        // Fallback to less accurate delta with DD's framework load time
+        _processStartTime = _frameworkLoadTime;
+    }
```

Comment:
do we know documented cases where `res != 0`?
this is quite a longshot but if `res` is practically always 0 and if we can find a way to execute code pre-main in Swift, we can call `sysctl` from Swift and remove this ObjC code. eventually, we can also remove `_Datadog_Private`

Reply:
As per the documentation it's only going to be 0 or -1:
```
Upon successful completion, the value 0 is	returned; otherwise the
     value -1 is returned and the global variable errno	is set to indicate the
     error.
```

In general, we should be able to call `sysctl()` at any time on the Swift side, but to get close to a relevant app launch time without the Objc `+(void)load` trick we should either ask for explicit calls from our users (which could be easily overlooked), or integrate this mechanism into the Datadog/singleton setup that is hopefully done as part of the `UIApplication.delegate` `application(...:didFinishLaunching...)` by our users.
An obvious downside of that is spreading the whole app launch time state across our SDK, along with potential discrepancies if our assumptions of Datadog's setup are wrong.